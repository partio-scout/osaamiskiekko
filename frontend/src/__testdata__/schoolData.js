export default [
  {
    "id": 8,
    "typed_id": "SCHOOL_8",
    "name_en": "school2",
    "name_fi": "school2",
    "name_sv": "school2",
    "description_en": null,
    "description_fi": null,
    "description_sv": null,
    "url": null,
    "logoUrl": null,
    "created_at": "2019-06-03T08:29:05.438Z",
    "updated_at": "2019-06-03T08:29:05.456Z",
    "academicdegrees": [
      {
        "id": 6,
        "identification": "school2-deg1",
        "name_en": "school2-deg1",
        "name_fi": "school2-deg1",
        "name_sv": "school2-deg1",
        "description_en": "school2-deg1",
        "description_fi": "school2-deg1",
        "description_sv": "school2-deg1",
        "url": null,
        "credits": null,
        "nqf": 2,
        "school": 8,
        "fieldofstudy": 2,
        "created_at": "2019-06-03T08:32:25.497Z",
        "updated_at": "2019-06-03T08:32:25.512Z"
      }
    ]
  },
  {
    "id": 9,
    "typed_id": "SCHOOL_9",
    "name_en": "school3",
    "name_fi": "school3",
    "name_sv": "school3",
    "description_en": null,
    "description_fi": null,
    "description_sv": null,
    "url": null,
    "logoUrl": null,
    "created_at": "2019-06-03T08:29:42.279Z",
    "updated_at": "2019-06-03T08:29:42.301Z",
    "academicdegrees": []
  },
  {
    "id": 7,
    "typed_id": "SCHOOL_7",
    "name_en": "school1  has a long name that doesn't fit on one row_sv",
    "name_fi": "school1 nimi ei mahdu yhdelle riville koska se on hirmu pitkä_fi",
    "name_sv": "school1  har ett långt namn som inte passar på en rad_sv",
    "description_en": "school1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._en",
    "description_fi": "school1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._fi",
    "description_sv": "school1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._sv",
    "url": null,
    "logoUrl": null,
    "created_at": "2019-06-03T08:28:57.373Z",
    "updated_at": "2019-06-03T08:39:35.828Z",
    "academicdegrees": [
      {
        "id": 4,
        "identification": "school1-deg2",
        "name_en": "school1-deg2",
        "name_fi": "school1-deg2",
        "name_sv": "school1-deg2",
        "description_en": "school1-deg2",
        "description_fi": "school1-deg2",
        "description_sv": "school1-deg2",
        "url": null,
        "credits": null,
        "nqf": 5,
        "school": 7,
        "fieldofstudy": 2,
        "created_at": "2019-06-03T08:31:36.135Z",
        "updated_at": "2019-06-03T08:39:35.832Z"
      },
      {
        "id": 5,
        "identification": "school1-deg3",
        "name_en": "school1-deg3",
        "name_fi": "school1-deg3",
        "name_sv": "school1-deg3",
        "description_en": "school1-deg3",
        "description_fi": "school1-deg3",
        "description_sv": "school1-deg3",
        "url": null,
        "credits": null,
        "nqf": 4,
        "school": 7,
        "fieldofstudy": 5,
        "created_at": "2019-06-03T08:31:57.018Z",
        "updated_at": "2019-06-03T08:39:35.832Z"
      },
      {
        "id": 3,
        "identification": "school1-deg1",
        "name_en": "school1-deg1_en",
        "name_fi": "school1-deg1_fi",
        "name_sv": "school1-deg1_sv",
        "description_en": "school1-deg1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._en",
        "description_fi": "school1-deg1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._fi",
        "description_sv": "school1-deg1 Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec enim lectus, luctus nec lorem id, laoreet vestibulum justo. Curabitur imperdiet ac magna eu ornare. Ut vehicula eleifend vestibulum. Mauris eget purus et libero finibus malesuada. Maecenas magna libero, pellentesque cursus varius sed, aliquam eu felis. Maecenas congue augue vehicula massa tincidunt tincidunt. Suspendisse potenti. Suspendisse vulputate nisi eget nunc interdum ultricies. Etiam urna velit, imperdiet eget sollicitudin sit amet, suscipit eget augue. Ut commodo enim nec dapibus porttitor. Donec id congue tortor, a ullamcorper metus. Vivamus vel eros vel orci auctor lobortis. Cras ac neque diam.\n\nVivamus luctus dignissim odio, sit amet tristique nisl bibendum ac. Quisque ullamcorper eros et hendrerit scelerisque. Pellentesque vehicula ex urna, ut tempus ante efficitur a. Aliquam metus nibh, vulputate a augue ac, auctor ullamcorper mi. Donec vitae facilisis ante, sagittis egestas mi. Nunc dignissim velit vitae ullamcorper suscipit. Sed posuere, est non tempor molestie, elit nibh consectetur metus, sit amet tincidunt ante nibh quis ex. Sed vestibulum at dolor vel pulvinar._sv",
        "url": null,
        "credits": 14,
        "nqf": 3,
        "school": 7,
        "fieldofstudy": 2,
        "created_at": "2019-06-03T08:31:16.758Z",
        "updated_at": "2019-06-03T08:39:35.832Z"
      }
    ]
  }
];
