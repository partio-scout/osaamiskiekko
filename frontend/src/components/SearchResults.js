import React from 'react';
import styled from 'styled-components';

const S = {};
S.ResultsDiv = styled.div`
  border-bottom: 1px solid #335362;
  padding: 10px 10px 10px 10px;
  :focus {
    outline: none;
    background-color: #E6E9EB;
  }

  p {
    color: #241C80;
    margin: 5px;
  }

  .type {
    font-size: 12px;	
    letter-spacing: 2px;
    line-height: 15px;
  }


  :last-child { border-bottom: none; }
  :hover {
   background-color: #E6E9EB;
   cursor: pointer;
  }
`;

S.ResultWrapper = styled.div`
  background-color: white;
  border-radius: 8px;
  box-shadow: 0 2px 8px 0 rgba(0,0,0,0.24);
  width: 475px;
  margin: auto;
  max-height: 238px;
  overflow: scroll;
`;


export default function SearchResults(props) {
  const { results, setSelection, globalState } = props;
  const renderResults = (item, index) =>
    <S.ResultsDiv 
      key={index} 
      onClick={() => setSelection(item)} 
      onKeyPress={() => setSelection(item)}
      tabIndex="0">
      <p>{item[`name_${globalState.language}`]}</p>
      { 
        item && item.type_fi ? 
        <p className="type">{item[`type_${globalState.language}`]}</p> : ''
      }
    </S.ResultsDiv>

  return (
    <S.ResultWrapper>
      {results.map(renderResults)}
    </S.ResultWrapper>
  )
}