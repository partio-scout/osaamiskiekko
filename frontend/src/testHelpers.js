import React from 'react';
import { IntlProvider, addLocaleData } from 'react-intl';
import { ThemeProvider } from 'styled-components';

import GlobalStateContext from './utils/GlobalStateContext';
import fi from 'react-intl/locale-data/fi';
import translations from './translations/translations';
import { defaultTheme } from './styles/Themes';

addLocaleData([...fi, translations]);

export default function renderWithProviders(component, globalState) {

  return (
    <GlobalStateContext.Provider value={globalState}>
      <ThemeProvider theme={defaultTheme}>
        <IntlProvider locale={'fi'} messages={translations['fi']}>
          {component}
        </IntlProvider>
      </ThemeProvider>
    </GlobalStateContext.Provider>
  );
};
