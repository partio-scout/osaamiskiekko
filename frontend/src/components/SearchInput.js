import React from 'react';
import styled from 'styled-components';
// import { DebounceInput } from 'react-debounce-input';
import { FormattedMessage } from 'react-intl';
import SearchResults from './SearchResults';

const S = {};
S.SearchWrapper = styled.div`
  margin: auto;
  position: relative;
  display: flex;
  flex-direction: column;
  margin-top: 30px;
  width: 80%;

label {
  height: 16px;	
  color: ${props => props.theme.colors.osaamisKiekkoBlue};
  font-size: 16px;	
  letter-spacing: 2.29px;
  line-height: 0px;
}

input {
  height: 32px;
  border: none;
  border-bottom: 2px solid ${props => props.theme.colors.osaamisKiekkoBlue};
  padding-left: 32px;
  outline-width: 0;
  outline: none;
  font-size: 20px;	
  line-height: 30px;
}

input::placeholder {
  color: ${props => props.theme.colors.osaamisKiekkoBlue};
  opacity: 1;
  font-size: 20px;	
  line-height: 30px;
  /* color: #202020; */
}

.fa-search { 
  position: absolute;
  left: 0px;
  top: 23px;
  font-size: 22px;
  color: ${props => props.theme.colors.osaamisKiekkoBlue};
}
`;

export default function SearchInput(props) {
  const { handleInput, inputValue, label, name, results, setSelection, showPreResults } = props;
  return (
      <S.SearchWrapper>
        <label htmlFor="search-input">
        <FormattedMessage id={label} />
        </label>
        <span className="fa fa-search"></span>
        <FormattedMessage id={label}>
          {msg => (
            <input
              minLength={2}
              // debounceTimeout={300}
              value={inputValue}
              type="text"
              placeholder={msg}
              name={name}
              onChange={e => handleInput(e.target.value)} 
              onClick={(e) => showPreResults(e.target.name)}
              onFocus={(e) => showPreResults(e.target.name)}
              autoComplete="off"
              />
          )}
        </FormattedMessage>
        <SearchResults
          results={results}
          setSelection={setSelection}
        />
      </S.SearchWrapper>
  )
}
