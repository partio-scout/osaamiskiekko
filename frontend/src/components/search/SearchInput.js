import React, { useState } from 'react';
import styled from 'styled-components';
import { FormattedMessage, injectIntl } from 'react-intl';
import _ from 'lodash';
import SearchResults from './SearchResults';

const S = {};
S.SearchWrapper = styled.div`
margin: auto;
position: relative;
display: flex;
flex-direction: column;

label {
  color: ${props => props.theme.colors.textColor};
  font-size: 16px;	
  letter-spacing: 2.29px;
  line-height: 20px;

  span {
    display: block;
    margin-bottom: 0.7em;
  }

  .input-area {
    position:relative;
  }
  
  input {
    box-sizing: border-box;
    height: 45px;
    width: 100%;
    border: none;
    border-bottom: 1px solid ${props => props.theme.colors.textColor};
    padding-left: 32px;
    outline-width: 0;
    outline: none;
    font-size: 20px;	
    line-height: 30px;
    font-weight: 300;
  }
  
  input::placeholder {
    color: ${props => props.theme.colors.textColor};
    opacity: 1;
    font-size: 20px;	
    line-height: 30px;
    font-style: italic;	
  }
  
  .fa-search { 
    position: absolute;
    width: 22px;
    font-size: 22px;
    line-height: 45px;
    color: ${props => props.theme.colors.textColor};
  }
}

@media only screen and (max-width: 767px) {
  label {
    font-size: 14px;
    letter-spacing: 2px;
    line-height: 15px;

    input, input::placeholder {
      font-size: 16px;
      line-height: 24px
    }
  
    fa-search: {
      font-size: 20px;
    }
  }

}
`;

export default injectIntl(function SearchInput(props) {
  const { handleInput, inputValue, labelKey, name, results, setSelection, showPreResults } = props;
  const [id] = useState(_.uniqueId());

  const mobileAutoScroll = () => {
    var elmnt = document.getElementById("search-box");
    if (window.innerWidth <= 767) {
      elmnt.scrollIntoView();
    }
  };

  return (
    <S.SearchWrapper name={`${name}-component`}>
      <label htmlFor={id}>
        <FormattedMessage id={labelKey} />
        <div className="input-area" role='button' onFocus={(e) => mobileAutoScroll()}> 
          <span className="fa fa-search" aria-hidden={true}></span>
          <input
            minLength={2}
            value={inputValue}
            type="text"
            placeholder={props.intl.formatMessage({id: 'search.placeholder', defaultMessage: 'Search...'})}
            name={name}
            onChange={e => handleInput(e.target.value)} 
            onClick={(e) => showPreResults(e.target.name)}
            onFocus={(e) => showPreResults(e.target.name)}
            autoComplete="off"
            id={id}
          />
        </div>
      </label>
      <SearchResults
        results={results}
        setSelection={setSelection}
      />
    </S.SearchWrapper>
  )
});
