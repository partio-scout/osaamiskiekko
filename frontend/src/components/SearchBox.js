import React, { useState }  from 'react';
import styled from 'styled-components';
import SearchInput from './SearchInput';
import { css } from '@emotion/core';
import { BarLoader } from 'react-spinners';
import getSchoolsAndOrganizations from '../api/GetSchoolsAndOrganizations';
import { useGlobalStateContext } from '../utils/GlobalStateContext';

const S = {};
S.SearchBox = styled.div`
  background-color: white;
  min-height: 125px;	
  width: 587px;
  background-color: #FFFFFF;
 	box-shadow: 0 2px 8px 0 rgba(0,0,0,0.24);
  border-radius: 8px;
  padding: 34px;
  margin: auto;

  .search-wrapper {
    margin-bottom: 35px;
  }

  @media only screen and (max-width: 945px) {
    margin-top: 0px;
  }

  @media only screen and (max-width: 800px) {
    width: 350px;
    margin-top: -20px;
  }
`;

const loadingSpinnerOverride = css`
    margin: auto;
    margin-top: 60px;
`;

export default function SearchBox(props) {
  const { showResults } = props;
  const globalState = useGlobalStateContext();
  const { data, isLoading } = getSchoolsAndOrganizations();

  const [inputSchoolOrOrganization, setinputSchoolOrOrganization] = useState('');
  const [schoolOrOrganizationFilter, setschoolOrOrganizationFilter] = useState([]);
  const [schoolOrOrganizationSelection, setschoolOrOrganizationSelection] = useState(null);

  const [inputTrainingValue, setinputTrainingValue] = useState('');
  const [competenceOrDegreeFilter, setcompetenceOrDegreeFilter] = useState([]);

  const noResults = {
    id: 1,
    error: true,
    name_fi: 'Tuloksia ei lÃ¶ytynyt',
    name_en: 'No search results',
    name_sv: "Inga resultat hittades"
  };

  const searchComparison = (item, searchValue) => 
    item.name_en.toUpperCase().includes(searchValue.toUpperCase()) ||
    item.name_fi.toUpperCase().includes(searchValue.toUpperCase()) ||
    item.name_sv.toUpperCase().includes(searchValue.toUpperCase());

  const filterSchoolOrOrganization = (searchValue) => {
    setinputSchoolOrOrganization(searchValue);
    if (searchValue) {
      setschoolOrOrganizationSelection(null);
      const results = data.filter(item => searchComparison(item, searchValue));
      results.length > 0 ?
      setschoolOrOrganizationFilter (results) :
      setschoolOrOrganizationFilter([noResults]);
      setinputTrainingValue('');
    } else {
      setschoolOrOrganizationFilter([]);
      setcompetenceOrDegreeFilter([]);
      setinputTrainingValue('');
      setschoolOrOrganizationSelection(null);
    }
  }

  const setDegreeOrCompetenceResults = (results) =>
    results.length > 0 ?
    setcompetenceOrDegreeFilter(results) :
    setcompetenceOrDegreeFilter([noResults]);

  const filterDegreesOrCompetences = (searchValue) => {
    setinputTrainingValue(searchValue);
    if (searchValue) {
      if (schoolOrOrganizationSelection.type_en === 'School') {
        const results = schoolOrOrganizationSelection.academicdegrees
          .filter(item => searchComparison(item, searchValue));
        setDegreeOrCompetenceResults(results);
      } else {
        const results = schoolOrOrganizationSelection.competences
          .filter(item => searchComparison(item, searchValue));
        setDegreeOrCompetenceResults(results);
      }
    } else {
      setcompetenceOrDegreeFilter([]);
    }
  }

  const getUserSelectionForSchoolOrAcademy = (selection) => {
    if (selection.error) return;
    setschoolOrOrganizationSelection(selection);
    setinputSchoolOrOrganization(selection[`name_${globalState.language}`]);
    setschoolOrOrganizationFilter([]);
  };

  const getUserSelectionForDegreeOrCompetence = (selection) => {
    if (selection.error) return;
    setinputTrainingValue(selection[`name_${globalState.language}`]);
    setcompetenceOrDegreeFilter([]);
    showResults(schoolOrOrganizationSelection, selection);
  }

  return (
    <S.SearchBox>
      {isLoading &&  
      <BarLoader
        css={loadingSpinnerOverride}
        sizeUnit={"px"}
        size={150}
        color={'#00283B'}
        loading={isLoading}
      />}
      {!isLoading &&
        <div className="search-wrapper">
          <SearchInput 
            handleInput={filterSchoolOrOrganization} 
            inputValue={inputSchoolOrOrganization} 
            label={'searchbox.label'} 
            name={'search-school'}
            results={schoolOrOrganizationFilter}
            setSelection={getUserSelectionForSchoolOrAcademy}
            />
        {schoolOrOrganizationSelection &&
          <SearchInput 
            handleInput={filterDegreesOrCompetences}
            inputValue={inputTrainingValue} 
            label={'searchbox.labelSecondary'}
            name={'search-education' }
            results={competenceOrDegreeFilter}
            setSelection={getUserSelectionForDegreeOrCompetence}
            />
          }
        </div>
      }
    </S.SearchBox>
  )
}
