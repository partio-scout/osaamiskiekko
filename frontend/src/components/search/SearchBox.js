import React, { useState, useEffect }  from 'react';
import styled from 'styled-components';
import SearchInput from './SearchInput';
import { css } from '@emotion/core';
import { BarLoader } from 'react-spinners';
import { useGlobalStateContext } from '../../utils/GlobalStateContext';
import OutsideClickHandler from 'react-outside-click-handler';

const S = {};
S.SearchBox = styled.div`
  background-color: #FFFFFF;
 	box-shadow: 0 2px 8px 0 rgba(0,0,0,0.24);
  border-radius: 8px;
  margin-top: -500px;
  z-index: 20;
  position: relative;
  margin-left: auto;
  margin-right: auto;
  width: 587px;
  display: flex;
  justify-content: center;
  align-items: start;
  padding: 32px 73px;

  .search-wrapper {
    width: 100%;

    > div {
      margin-top: 2.5em;

      &:first-child {
        margin-top: 0;
      }
    }
  }

  @media only screen and (max-width: 767px) {
    width: 85%;
    padding: 16px;
  }
`;

const loadingSpinnerOverride = css`
    margin: auto;
    margin-top: 60px;
`;

export default function SearchBox(props) {
  const { showResults, data, isLoading, clearResults } = props;
  const globalState = useGlobalStateContext();

  const {selectedInstitution, selectedTraining} = globalState;

  const [institutionInput, setInstitutionInput] = useState('');
  const [institutionFilter, setInstitutionFilter] = useState([]);
  const [institutionSelection, setInstitutionSelection] = useState(null);
  const [trainingInput, setTrainingInput] = useState('');
  const [trainingFilter, setTrainingFilter] = useState([]);

  const updateInputsWithInstitution = (institution) => {
    setInstitutionSelection(institution);
    setInstitutionInput(institution[`name_${globalState.language}`]);
    setInstitutionFilter([]);
    setTrainingFilter([]);
    setTrainingInput('');
    clearResults();
  }

  const updateInputsWithTraining = (training) => {
    setTrainingInput(training[`name_${globalState.language}`]);
    setTrainingFilter([]);
  }
  
  useEffect(() => {
    if (selectedInstitution && selectedTraining) {
      updateInputsWithInstitution(selectedInstitution);
      updateInputsWithTraining(selectedTraining);
      showResults(selectedInstitution, selectedTraining);
    }
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  const noResults = {
    id: 1,
    error: true,
    name_fi: 'Tuloksia ei lÃ¶ytynyt',
    name_en: 'No search results',
    name_sv: "Inga resultat hittades"
  };

  const searchComparison = (item, searchValue) => 
    item.name_en.toUpperCase().includes(searchValue.toUpperCase()) ||
    item.name_fi.toUpperCase().includes(searchValue.toUpperCase()) ||
    item.name_sv.toUpperCase().includes(searchValue.toUpperCase());

  const showPreResults = (e) => {
    if (e === 'search-school') setInstitutionFilter(data);
    if (e === 'search-education') {
      if (institutionSelection.type_en === 'School') {
        setDegreeOrCompetenceResults(institutionSelection.academicdegrees);
      } else {
        setDegreeOrCompetenceResults(institutionSelection.competences);
      }
    }
  };

  const filterSchoolOrOrganization = (searchValue) => {
    setInstitutionInput(searchValue);
    if (searchValue) {
      setInstitutionSelection(null);
      const results = data.filter(item => searchComparison(item, searchValue));
      results.length > 0 ?
      setInstitutionFilter (results) :
      setInstitutionFilter([noResults]);
      setTrainingInput('');
    } else {
      setInstitutionFilter([]);
      setTrainingFilter([]);
      setTrainingInput('');
      setInstitutionSelection(null);
    }
  }

  const setDegreeOrCompetenceResults = (results) =>
    results.length > 0 ?
    setTrainingFilter(results) :
    setTrainingFilter([noResults]);

  const filterDegreesOrCompetences = (searchValue) => {
    setTrainingInput(searchValue);
    if (searchValue) {
      if (institutionSelection.type_en === 'School') {
        const results = institutionSelection.academicdegrees
          .filter(item => searchComparison(item, searchValue));
        setDegreeOrCompetenceResults(results);
      } else {
        const results = institutionSelection.competences
          .filter(item => searchComparison(item, searchValue));
        setDegreeOrCompetenceResults(results);
      }
    } else {
      setTrainingFilter([]);
    }
  }
  
  const userInstitutionSelection = (selection) => {
    if (selection.error) return;

    globalState.selectedInstitution = selection;
    updateInputsWithInstitution(selection);
  };

  const userTrainingSelection = (selection) => {
    if (selection.error) return;

    globalState.selectedTraining = selection;
    updateInputsWithTraining(selection);
    showResults(institutionSelection, selection);
  }

  return (
    <S.SearchBox id="search-box">
      {isLoading &&  
      <BarLoader
        css={loadingSpinnerOverride}
        sizeUnit={"px"}
        size={150}
        color={'#00283B'}
        loading={isLoading}
      />}
      {!isLoading &&
        <div className="search-wrapper">
          <OutsideClickHandler
            onOutsideClick={() => {
              setInstitutionFilter([]);
            }}
          >
          <SearchInput 
            handleInput={filterSchoolOrOrganization} 
            inputValue={institutionInput} 
            labelKey='search.label' 
            name='search-school'
            className='search-school'
            results={institutionFilter}
            setSelection={userInstitutionSelection}
            showPreResults={showPreResults}
            />
          </OutsideClickHandler>
            {institutionSelection &&
              <OutsideClickHandler
                onOutsideClick={() => {
                  setTrainingFilter([]);
                }}
              >
            <SearchInput
              handleInput={filterDegreesOrCompetences}
              inputValue={trainingInput} 
              labelKey='search.labelSecondary'
              name='search-education' 
              className='search-education'
              results={trainingFilter}
              setSelection={userTrainingSelection}
              showPreResults={showPreResults}
            />
          </OutsideClickHandler>
            }
        </div>
      }
    </S.SearchBox>
  )
}
