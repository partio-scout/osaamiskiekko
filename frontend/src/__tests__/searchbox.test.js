import React from 'react';
import {render, shallow, mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import SearchBox from '../components/SearchBox';
import SearchResults from '../components/SearchResults';
import SearchInput from '../components/SearchInput';
import * as testHelper from '../testHelpers';
import schoolData from '../__testdata__/schoolData'

configure({adapter: new Adapter()});

describe('search results', () => {
  test('shows results', () => {
    const wrapper = mount(testHelper.wrapWithProviders(<SearchResults results={schoolData} />));
    expect(wrapper.find('li.search-result-item').length).toBe(schoolData.length);
  });

  test('calls selection handler', () => {
    const setSelectionCallback = jest.fn();
    const wrapper = mount(testHelper.wrapWithProviders(<SearchResults results={schoolData} setSelection={setSelectionCallback} />));
    wrapper.find('li.search-result-item').first().simulate('click');
    expect(setSelectionCallback.mock.calls.length).toBe(1);
  });
});


describe('search input', () => {
  test('shows results', () => {
    const showPreResultsCallback = jest.fn();
    const wrapper = mount(testHelper.wrapWithProviders(<SearchInput label='searchbox.label' results={schoolData} showPreResults={showPreResultsCallback} />));
    wrapper.find('input').simulate('click');
    expect(wrapper.find('li.search-result-item').length).toBe(schoolData.length);
  });

  test('calls showPreResults', () => {
    const showPreResultsCallback = jest.fn();
    const wrapper = mount(testHelper.wrapWithProviders(<SearchInput label='searchbox.label' results={[]} showPreResults={showPreResultsCallback} />));
    wrapper.find('input').simulate('click');
    wrapper.find('input').simulate('focus');
    expect(showPreResultsCallback.mock.calls.length).toBe(2);
  });

  test('calls handleInput', () => {
    const showPreResultsCallback = jest.fn();
    const handleInputCallback = jest.fn();
    const wrapper = mount(testHelper.wrapWithProviders(
      <SearchInput 
        label='searchbox.label' 
        results={schoolData} 
        showPreResults={showPreResultsCallback}
        handleInput={handleInputCallback}
      />));
    wrapper.find('input').simulate('change', { target: { value: 'Partio' }});
    wrapper.find('input').simulate('change', { target: { value: 'Doesntexist' }});
    expect(handleInputCallback.mock.calls.length).toBe(2);
  });

  test('calls selection handler', () => {
    const setSelectionCallback = jest.fn();
    const wrapper = mount(testHelper.wrapWithProviders(<SearchResults results={schoolData} setSelection={setSelectionCallback} />));
    wrapper.find('li.search-result-item').first().simulate('click');
    expect(setSelectionCallback.mock.calls.length).toBe(1);
  });
});

describe('search box', () => {
  test('shows results', () => {
    const wrapper = mount(testHelper.wrapWithProviders(<SearchBox data={schoolData} isLoading={false} />));
    wrapper.find('.search-school input').simulate('click');
    expect(wrapper.find('li.search-result-item').length).toBe(schoolData.length);
  });

  test('filters results', () => {
    const wrapper = mount(testHelper.wrapWithProviders(<SearchBox data={schoolData} isLoading={false} />));
    wrapper.find('.search-school input').simulate('click');
    expect(wrapper.find('li.search-result-item').length).toBe(schoolData.length);
    wrapper.find('input').simulate('change', { target: { value: 'Partio' }});
    expect(wrapper.find('li.search-result-item').length).toBe(1);
  });

});
