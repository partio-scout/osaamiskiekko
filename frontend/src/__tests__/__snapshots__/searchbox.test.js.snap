// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox snapshot matches 1`] = `
<div
  className="sc-bxivhb SMrtp"
>
  <div
    className="sc-htpNat dmPsyL"
  >
    <label
      htmlFor="search-input"
    >
      <span>
        Valitse järjestö tai oppilaitos
      </span>
    </label>
    <span
      className="fa fa-search"
    />
    <input
      name="search-input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Valitse järjestö tai oppilaitos"
      type="text"
      value=""
    />
  </div>
  <div
    className="sc-bwzfXH jsMlNV"
  />
  <div
    className="sc-htpNat dmPsyL"
  >
    <label
      htmlFor="search-input"
    >
      <span>
        Valitse koulutus
      </span>
    </label>
    <span
      className="fa fa-search"
    />
    <input
      name="search-input"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Valitse koulutus"
      type="text"
      value=""
    />
  </div>
  <button>
    Näytä tulokset
  </button>
</div>
`;
